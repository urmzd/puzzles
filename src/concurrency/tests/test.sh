#!/bin/bash

do_diff() {
  # $1 expected
  # $2 is produced
  # $3 Test name

  echo -n "$3 test "
  diff $1 $2 > /dev/null
  if [ $? != 0 ]; then
    echo FAILED
    echo ======
    echo Expected Output:
    echo ++++++++++++++++
    cat $1
    echo =====================
    echo Actual Output:
    echo ++++++++++++++++
    cat $2
    exit 1
  else
    echo PASSED
  fi
}


echo ======================================================
echo ====================== TEST $1 ========================
echo ======================================================
ACTUAL=tests/test.$1.out
ACT_BASIC=tests/test.$1.act_bsc
ACT_RECV=tests/test.$1.act_recv
ACT_ONLY=tests/test.$1.act_only
ACTUAL2=tests/test.$1.out2
ACT2_RECV=tests/test.$1.act2_recv
EXPECT=tests/test.$1.expected
EXP_BASIC=tests/test.$1.exp_bsc
EXP_RECV=tests/test.$1.exp_recv
EXP_ONLY=tests/test.$1.exp_only

CFG=`cat tests/test.$1.cfg`
DELAY=tests/delay.py
INPUT=tests/test.$1.in

if which timeout > /dev/null; then
  TIMEOUT="timeout 10"
  echo "##" Using a timeout of 10 seconds
  echo ======================================================
else
  TIMEOUT=""
fi

cat tests/test.$1.desc
echo ======================================================
cat $INPUT | python3 $DELAY $CFG | $TIMEOUT ./miner/miner > $ACTUAL
EXIT=${PIPESTATUS[2]}
if [ $EXIT == 0 ]; then 
  grep -v ^Thread $ACTUAL > $ACT_RECV
  grep -v ^Thread $EXPECT > $EXP_RECV
  grep -v ^Received $ACT_RECV > $ACT_BASIC
  grep -v ^Received $EXP_RECV > $EXP_BASIC
  do_diff $EXP_BASIC $ACT_BASIC "Basic functionality"

  if grep BASIC tests/test.$1.cfg > /dev/null; then
    grep ^Received $ACTUAL > $ACT_ONLY
    grep ^Received $EXPECT > $EXP_ONLY
    
    do_diff $EXP_ONLY $ACT_ONLY "Receive statement ordering should be the same"
  elif grep RANDOM tests/test.$1.cfg > /dev/null; then
    cat $INPUT | python3 $DELAY $CFG | $TIMEOUT ./miner/miner > $ACTUAL2
    grep -v ^Thread $ACTUAL2 > $ACT2_RECV

    echo Input has been randomly delayed to create random behaviour
    echo Output from multiple runs should be different 
    diff $ACT_RECV $ACT2_RECV > /dev/null
    if [ $? == 0 ]; then
      echo Output from multiple runs is the same: test FAILED
      exit 1
    else
      echo Output from multiple runs is different: test PASSED
    fi
  else
    do_diff $EXP_RECV $ACT_RECV "Receive ordering should be the same"
  fi
elif [ $EXIT == 124 ]; then
  echo TIMEOUT
  exit 1
else 
  echo Abnormal program termination: the program crashed
  echo Exit code $?
  exit 1
fi
