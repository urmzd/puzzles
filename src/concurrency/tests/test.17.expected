Received event TRX with ID 11235
Received event TRX with ID 11236
Adding transaction: 11235 Al_Alpha Beth_Beta 150000 40
Adding transaction: 11236 Cat_Charlie Dov_Delta 150000 20
Received event TRX with ID 11237
Received event TRX with ID 11238
Adding transaction: 11237 Eve_Echo Fin_Foxtrot 150000 25
Adding transaction: 11238 Guy_Golf Hana_Hotel 150000 40
Received event TRX with ID 11239
Received event TRX with ID 11240
Adding transaction: 11239 Iona_Indigo Jeff_Juliet 150000 20
Adding transaction: 11240 Kit_Kilo Liam_Lima 150000 25
Received event TRX with ID 11241
Adding transaction: 11241 Mona_Mike Nick_November 150000 40
Received event TRX with ID 11242
Received event TRX with ID 11243
Adding transaction: 11242 Otto_Oscar Peta_Papa 150000 20
Adding transaction: 11243 Qu_Quebec Reya_Romeo 150000 25
Received event TRX with ID 23581
Adding transaction: 23581 Alice_25 Bob_283 705106 751
Received event TRX with ID 23582
Adding transaction: 23582 Alice_710 Bob_459 5346392 989
Received event TRX with ID 23583
Received event TRX with ID 23584
Adding transaction: 23583 Alice_548 Bob_19 1272463 993
Adding transaction: 23584 Alice_65 Bob_906 9343697 318
Received event TRX with ID 23585
Adding transaction: 23585 Alice_609 Bob_862 2819973 510
Received event TRX with ID 23586
Adding transaction: 23586 Alice_256 Bob_300 1281553 205
Received event TRX with ID 23587
Received event TRX with ID 23588
Received event TRX with ID 23589
Adding transaction: 23587 Alice_219 Bob_73 8439328 833
Adding transaction: 23588 Alice_41 Bob_380 7504293 892
Adding transaction: 23589 Alice_679 Bob_837 3064006 136
Received event TRX with ID 23590
Adding transaction: 23590 Alice_957 Bob_916 4317179 441
Received event TRX with ID 23591
Adding transaction: 23591 Alice_445 Bob_737 5404554 313
Received event TRX with ID 23592
Received event TRX with ID 23593
Adding transaction: 23592 Alice_282 Bob_626 5713910 312
Adding transaction: 23593 Alice_192 Bob_146 4040219 977
Received event TRX with ID 23594
Adding transaction: 23594 Alice_603 Bob_58 6826071 267
Received event TRX with ID 23595
Adding transaction: 23595 Alice_536 Bob_923 2683967 199
Received event TRX with ID 23596
Adding transaction: 23596 Alice_353 Bob_740 8125918 416
Received event TRX with ID 23597
Adding transaction: 23597 Alice_951 Bob_598 9994216 683
Received event TRX with ID 23598
Adding transaction: 23598 Alice_556 Bob_141 5323762 983
Received event TRX with ID 23599
Received event TRX with ID 23600
Received event TRX with ID 23601
Received event TRX with ID 23602
Adding transaction: 23599 Alice_926 Bob_455 6496146 86
Adding transaction: 23600 Alice_755 Bob_403 2575677 478
Adding transaction: 23601 Alice_605 Bob_752 2412567 533
Adding transaction: 23602 Alice_320 Bob_132 7640399 984
Received event TRX with ID 23603
Received event TRX with ID 23604
Adding transaction: 23603 Alice_967 Bob_673 2371813 191
Adding transaction: 23604 Alice_261 Bob_226 8757863 769
Received event TRX with ID 23605
Adding transaction: 23605 Alice_190 Bob_505 8997535 575
Received event TRX with ID 23606
Received event TRX with ID 23607
Received event TRX with ID 23608
Received event TRX with ID 23609
Adding transaction: 23606 Alice_381 Bob_211 6268875 267
Adding transaction: 23607 Alice_55 Bob_572 279178 630
Adding transaction: 23608 Alice_82 Bob_437 5932332 557
Adding transaction: 23609 Alice_159 Bob_663 8068773 709
Received event TRX with ID 23610
Received event TRX with ID 23611
Adding transaction: 23610 Alice_757 Bob_826 1790414 82
Adding transaction: 23611 Alice_159 Bob_614 1575759 575
Received event TRX with ID 23612
Adding transaction: 23612 Alice_995 Bob_633 3720510 507
Received event TRX with ID 23613
Adding transaction: 23613 Alice_898 Bob_500 8527378 984
Received event TRX with ID 23614
Received event TRX with ID 23615
Received event TRX with ID 23616
Adding transaction: 23614 Alice_755 Bob_749 7337492 194
Adding transaction: 23615 Alice_225 Bob_479 6791212 383
Adding transaction: 23616 Alice_901 Bob_7 6644670 418
Received event TRX with ID 23617
Adding transaction: 23617 Alice_406 Bob_988 8760577 544
Received event TRX with ID 23618
Received event TRX with ID 23619
Received event TRX with ID 23620
Adding transaction: 23618 Alice_981 Bob_759 5698765 198
Adding transaction: 23619 Alice_124 Bob_515 9325067 430
Adding transaction: 23620 Alice_494 Bob_284 746468 138
Received event TRX with ID 23621
Received event TRX with ID 23622
Adding transaction: 23621 Alice_905 Bob_104 9631652 256
Adding transaction: 23622 Alice_386 Bob_180 861886 946
Received event TRX with ID 23623
Adding transaction: 23623 Alice_166 Bob_686 8683562 774
Received event TRX with ID 23624
Adding transaction: 23624 Alice_901 Bob_987 5867398 604
Received event TRX with ID 23625
Adding transaction: 23625 Alice_515 Bob_12 6520129 360
Received event TRX with ID 23626
Adding transaction: 23626 Alice_287 Bob_852 3418434 752
Received event TRX with ID 23627
Received event TRX with ID 23628
Adding transaction: 23627 Alice_376 Bob_955 3857856 939
Adding transaction: 23628 Alice_957 Bob_543 358279 607
Received event TRX with ID 23629
Adding transaction: 23629 Alice_196 Bob_703 6457746 150
Received event TRX with ID 23630
Adding transaction: 23630 Alice_709 Bob_802 4499054 564
Received event TRX with ID 23631
Received event TRX with ID 23632
Adding transaction: 23631 Alice_962 Bob_880 2959400 514
Adding transaction: 23632 Alice_760 Bob_492 5684360 272
Received event TRX with ID 23633
Received event TRX with ID 23634
Received event TRX with ID 23635
Received event TRX with ID 23636
Received event TRX with ID 23637
Adding transaction: 23633 Alice_539 Bob_92 2048208 290
Received event TRX with ID 23638
Adding transaction: 23634 Alice_352 Bob_342 9407464 603
Adding transaction: 23635 Alice_259 Bob_888 698819 652
Adding transaction: 23636 Alice_27 Bob_150 8939561 958
Adding transaction: 23637 Alice_612 Bob_265 322697 947
Adding transaction: 23638 Alice_233 Bob_600 6225988 39
Received event TRX with ID 23639
Adding transaction: 23639 Alice_180 Bob_451 438396 603
Received event TRX with ID 23640
Adding transaction: 23640 Alice_126 Bob_453 3630462 745
Received event TRX with ID 23641
Adding transaction: 23641 Alice_302 Bob_763 883875 634
Received event TRX with ID 23642
Adding transaction: 23642 Alice_396 Bob_195 2560502 478
Received event TRX with ID 23643
Received event TRX with ID 23644
Adding transaction: 23643 Alice_767 Bob_746 4587301 339
Adding transaction: 23644 Alice_146 Bob_266 6529085 982
Received event TRX with ID 23645
Adding transaction: 23645 Alice_895 Bob_262 3926500 541
Received event TRX with ID 23646
Adding transaction: 23646 Alice_931 Bob_230 2393155 293
Received event TRX with ID 23647
Adding transaction: 23647 Alice_145 Bob_15 4017767 508
Received event TRX with ID 23648
Received event TRX with ID 23649
Adding transaction: 23648 Alice_389 Bob_609 3613138 210
Adding transaction: 23649 Alice_107 Bob_865 5781800 80
Received event TRX with ID 23650
Adding transaction: 23650 Alice_697 Bob_618 550217 445
Received event TRX with ID 23651
Received event TRX with ID 23652
Adding transaction: 23651 Alice_667 Bob_500 5903392 366
Adding transaction: 23652 Alice_502 Bob_777 3929724 86
Received event TRX with ID 23653
Received event TRX with ID 23654
Adding transaction: 23653 Alice_490 Bob_45 4567699 350
Adding transaction: 23654 Alice_88 Bob_446 3228103 162
Received event TRX with ID 23655
Received event TRX with ID 23656
Adding transaction: 23655 Alice_251 Bob_899 3524757 177
Adding transaction: 23656 Alice_387 Bob_774 9677443 494
Received event TRX with ID 23657
Adding transaction: 23657 Alice_539 Bob_17 6369128 72
Received event TRX with ID 23658
Adding transaction: 23658 Alice_956 Bob_884 9731275 958
Received event TRX with ID 23659
Received event TRX with ID 23660
Adding transaction: 23659 Alice_492 Bob_930 3477798 885
Adding transaction: 23660 Alice_73 Bob_593 7327291 790
Received event TRX with ID 23661
Adding transaction: 23661 Alice_409 Bob_670 8611706 636
Received event TRX with ID 23662
Received event TRX with ID 23663
Received event TRX with ID 23664
Adding transaction: 23662 Alice_80 Bob_98 9211801 507
Adding transaction: 23663 Alice_33 Bob_36 677522 280
Adding transaction: 23664 Alice_196 Bob_524 4510833 61
Received event TRX with ID 23665
Received event TRX with ID 23666
Received event TRX with ID 23667
Received event TRX with ID 23668
Received event TRX with ID 23669
Adding transaction: 23665 Alice_98 Bob_206 9311992 576
Adding transaction: 23666 Alice_399 Bob_253 7304649 101
Adding transaction: 23667 Alice_404 Bob_432 9169719 28
Adding transaction: 23668 Alice_0 Bob_34 1433090 710
Adding transaction: 23669 Alice_27 Bob_290 546915 418
Received event TRX with ID 23670
Received event TRX with ID 23671
Adding transaction: 23670 Alice_25 Bob_687 4460250 433
Adding transaction: 23671 Alice_722 Bob_894 543955 522
Received event TRX with ID 23672
Adding transaction: 23672 Alice_904 Bob_533 1693820 187
Received event TRX with ID 23673
Adding transaction: 23673 Alice_989 Bob_11 813372 944
Received event TRX with ID 23674
Adding transaction: 23674 Alice_866 Bob_231 2469028 44
Received event TRX with ID 23675
Adding transaction: 23675 Alice_515 Bob_16 5744227 188
Received event TRX with ID 23676
Received event TRX with ID 23677
Received event TRX with ID 23678
Adding transaction: 23676 Alice_440 Bob_793 3009268 603
Adding transaction: 23677 Alice_942 Bob_361 3690673 519
Adding transaction: 23678 Alice_899 Bob_508 2161599 198
Received event TRX with ID 23679
Received event TRX with ID 23680
Received event MINE
Adding transaction: 23679 Alice_866 Bob_594 5988232 505
Adding transaction: 23680 Alice_369 Bob_390 118372 532
Block mined: 1 0 0x00000000 8
23581 Alice_25 Bob_283 705106 751
23582 Alice_710 Bob_459 5346392 989
23583 Alice_548 Bob_19 1272463 993
23584 Alice_65 Bob_906 9343697 318
23585 Alice_609 Bob_862 2819973 510
23587 Alice_219 Bob_73 8439328 833
23588 Alice_41 Bob_380 7504293 892
23668 Alice_0 Bob_34 1433090 710
Thread 0 checking nonce 0x00000000
Thread 0 checking nonce 0x00000008
Thread 0 checking nonce 0x00000010
Thread 0 checking nonce 0x00000018
Thread 0 checking nonce 0x00000020
Thread 0 checking nonce 0x00000028
Thread 0 checking nonce 0x00000030
Thread 0 checking nonce 0x00000038
Thread 0 checking nonce 0x00000040
Thread 0 checking nonce 0x00000048
Thread 0 checking nonce 0x00000050
Thread 1 checking nonce 0x00000001
Thread 1 checking nonce 0x00000009
Thread 1 checking nonce 0x00000011
Thread 2 checking nonce 0x00000002
Thread 2 checking nonce 0x0000000a
Thread 2 checking nonce 0x00000012
Thread 2 checking nonce 0x0000001a
Thread 2 checking nonce 0x00000022
Thread 2 checking nonce 0x0000002a
Thread 2 checking nonce 0x00000032
Thread 2 checking nonce 0x0000003a
Thread 2 checking nonce 0x00000042
Thread 2 checking nonce 0x0000004a
Thread 2 checking nonce 0x00000052
Thread 2 checking nonce 0x0000005a
Thread 2 checking nonce 0x00000062
Thread 2 checking nonce 0x0000006a
Thread 2 checking nonce 0x00000072
Thread 2 checking nonce 0x0000007a
Thread 2 checking nonce 0x00000082
Thread 2 checking nonce 0x0000008a
Thread 2 checking nonce 0x00000092
Thread 2 checking nonce 0x0000009a
Thread 2 checking nonce 0x000000a2
Thread 0 checking nonce 0x00000058
Thread 4 checking nonce 0x00000004
Thread 0 checking nonce 0x00000060
Thread 0 checking nonce 0x00000068
Thread 0 checking nonce 0x00000070
Thread 0 checking nonce 0x00000078
Thread 0 checking nonce 0x00000080
Thread 0 checking nonce 0x00000088
Thread 2 checking nonce 0x000000aa
Thread 2 checking nonce 0x000000b2
Thread 1 checking nonce 0x00000019
Thread 1 checking nonce 0x00000021
Thread 1 checking nonce 0x00000029
Thread 4 checking nonce 0x0000000c
Thread 2 checking nonce 0x000000ba
Thread 4 checking nonce 0x00000014
Thread 2 checking nonce 0x000000c2
Thread 4 checking nonce 0x0000001c
Thread 4 checking nonce 0x00000024
Thread 6 checking nonce 0x00000006
Thread 2 checking nonce 0x000000ca
Thread 6 checking nonce 0x0000000e
Thread 5 checking nonce 0x00000005
Thread 5 checking nonce 0x0000000d
Thread 5 checking nonce 0x00000015
Thread 5 checking nonce 0x0000001d
Thread 5 checking nonce 0x00000025
Thread 5 checking nonce 0x0000002d
Thread 5 checking nonce 0x00000035
Thread 5 checking nonce 0x0000003d
Thread 5 checking nonce 0x00000045
Thread 5 checking nonce 0x0000004d
Thread 5 checking nonce 0x00000055
Thread 5 checking nonce 0x0000005d
Thread 5 checking nonce 0x00000065
Thread 2 checking nonce 0x000000d2
Thread 2 checking nonce 0x000000da
Thread 2 checking nonce 0x000000e2
Thread 4 checking nonce 0x0000002c
Thread 3 checking nonce 0x00000003
Thread 3 checking nonce 0x0000000b
Thread 3 checking nonce 0x00000013
Thread 3 checking nonce 0x0000001b
Thread 7 checking nonce 0x00000007
Thread 7 checking nonce 0x0000000f
Thread 7 checking nonce 0x00000017
Thread 5 checking nonce 0x0000006d
Thread 0 checking nonce 0x00000090
Thread 5 checking nonce 0x00000075
Thread 0 checking nonce 0x00000098
Thread 5 checking nonce 0x0000007d
Thread 0 checking nonce 0x000000a0
Thread 0 checking nonce 0x000000a8
Thread 0 checking nonce 0x000000b0
Thread 0 checking nonce 0x000000b8
Thread 0 checking nonce 0x000000c0
Thread 0 checking nonce 0x000000c8
Thread 0 checking nonce 0x000000d0
Thread 3 checking nonce 0x00000023
Thread 3 checking nonce 0x0000002b
Thread 6 checking nonce 0x00000016
Thread 5 checking nonce 0x00000085
Thread 4 checking nonce 0x00000034
Thread 4 checking nonce 0x0000003c
Thread 1 checking nonce 0x00000031
Thread 4 checking nonce 0x00000044
Thread 7 checking nonce 0x0000001f
Thread 7 checking nonce 0x00000027
Thread 0 checking nonce 0x000000d8
Thread 1 checking nonce 0x00000039
Thread 0 checking nonce 0x000000e0
Thread 1 checking nonce 0x00000041
Thread 0 checking nonce 0x000000e8
Thread 1 checking nonce 0x00000049
Thread 7 checking nonce 0x0000002f
Thread 7 checking nonce 0x00000037
Thread 4 checking nonce 0x0000004c
Thread 7 checking nonce 0x0000003f
Thread 4 checking nonce 0x00000054
Thread 5 checking nonce 0x0000008d
Thread 5 checking nonce 0x00000095
Thread 5 checking nonce 0x0000009d
Thread 5 checking nonce 0x000000a5
Thread 5 checking nonce 0x000000ad
Thread 7 checking nonce 0x00000047
Thread 3 checking nonce 0x00000033
Thread 7 checking nonce 0x0000004f
Thread 7 checking nonce 0x00000057
Thread 7 checking nonce 0x0000005f
Thread 7 checking nonce 0x00000067
Thread 7 checking nonce 0x0000006f
Thread 7 checking nonce 0x00000077
Thread 7 checking nonce 0x0000007f
Thread 7 checking nonce 0x00000087
Thread 7 checking nonce 0x0000008f
Thread 7 checking nonce 0x00000097
Thread 7 checking nonce 0x0000009f
Thread 4 checking nonce 0x0000005c
Thread 1 checking nonce 0x00000051
Thread 1 checking nonce 0x00000059
Thread 1 checking nonce 0x00000061
Thread 1 checking nonce 0x00000069
Thread 3 checking nonce 0x0000003b
Thread 7 checking nonce 0x000000a7
Thread 3 checking nonce 0x00000043
Thread 7 checking nonce 0x000000af
Thread 7 checking nonce 0x000000b7
Thread 4 checking nonce 0x00000064
Thread 4 checking nonce 0x0000006c
Thread 1 checking nonce 0x00000071
Thread 1 checking nonce 0x00000079
Thread 5 checking nonce 0x000000b5
Thread 5 checking nonce 0x000000bd
Thread 5 checking nonce 0x000000c5
Thread 1 checking nonce 0x00000081
Thread 1 checking nonce 0x00000089
Thread 6 checking nonce 0x0000001e
Thread 4 checking nonce 0x00000074
Thread 4 checking nonce 0x0000007c
Thread 4 checking nonce 0x00000084
Thread 3 checking nonce 0x0000004b
Thread 4 checking nonce 0x0000008c
Thread 4 checking nonce 0x00000094
Thread 4 checking nonce 0x0000009c
Thread 4 checking nonce 0x000000a4
Thread 4 checking nonce 0x000000ac
Thread 4 checking nonce 0x000000b4
Thread 4 checking nonce 0x000000bc
Thread 4 checking nonce 0x000000c4
Thread 4 checking nonce 0x000000cc
Thread 3 checking nonce 0x00000053
Thread 3 checking nonce 0x0000005b
Thread 6 checking nonce 0x00000026
Thread 3 checking nonce 0x00000063
Thread 3 checking nonce 0x0000006b
Thread 3 checking nonce 0x00000073
Thread 3 checking nonce 0x0000007b
Thread 3 checking nonce 0x00000083
Thread 3 checking nonce 0x0000008b
Thread 3 checking nonce 0x00000093
Thread 3 checking nonce 0x0000009b
Thread 3 checking nonce 0x000000a3
Thread 6 checking nonce 0x0000002e
Thread 6 checking nonce 0x00000036
Thread 7 checking nonce 0x000000bf
Thread 3 checking nonce 0x000000ab
Thread 3 checking nonce 0x000000b3
Thread 7 checking nonce 0x000000c7
Thread 3 checking nonce 0x000000bb
Thread 3 checking nonce 0x000000c3
Thread 3 checking nonce 0x000000cb
Thread 1 checking nonce 0x00000091
Thread 1 checking nonce 0x00000099
Thread 1 checking nonce 0x000000a1
Thread 1 checking nonce 0x000000a9
Thread 1 checking nonce 0x000000b1
Thread 1 checking nonce 0x000000b9
Thread 1 checking nonce 0x000000c1
Thread 1 checking nonce 0x000000c9
Thread 6 checking nonce 0x0000003e
Thread 6 checking nonce 0x00000046
Thread 6 checking nonce 0x0000004e
Thread 6 checking nonce 0x00000056
Thread 6 checking nonce 0x0000005e
Thread 6 checking nonce 0x00000066
Thread 6 checking nonce 0x0000006e
Thread 6 checking nonce 0x00000076
Thread 6 checking nonce 0x0000007e
Thread 6 checking nonce 0x00000086
Thread 6 checking nonce 0x0000008e
Thread 6 checking nonce 0x00000096
Thread 6 checking nonce 0x0000009e
Thread 6 checking nonce 0x000000a6
Thread 6 checking nonce 0x000000ae
Thread 6 checking nonce 0x000000b6
Thread 6 checking nonce 0x000000be
Thread 6 checking nonce 0x000000c6
0x000000c5 0x00e7f37c
Received event END
